python/ray/data/_internal/planner/plan_udf_map_op.py

```diff
@@ -606,3 +608,194 @@ def _create_map_transformer_for_block_based_map_op(
     ]
     return MapTransformer(transform_fns, init_fn=init_fn)
+
+
+_SENTINEL = object()
+
+T = TypeVar("T")
+U = TypeVar("U")
+
+
+def _generate_transform_fn_for_async_map(
+    fn: UserDefinedFunction,
+    validate_fn: Callable,
+    *,
+    max_concurrency: int,
+    is_flat_map: bool = False,
+) -> MapTransformCallable:
+    assert max_concurrency > 0, "Max concurrency must be positive"
+
+    if inspect.isasyncgenfunction(fn):
+
+        async def _apply_udf(item: T) -> List[U]:
+            gen = fn(item)
+            # NOTE: Async generator is unrolled inside the task to maintain
+            #       requested concurrency level (`max_concurrent_batches`)
+            return [out async for out in gen]
+
+    elif inspect.iscoroutinefunction(fn):
+
+        async def _apply_udf(item: T) -> List[U]:
+            res = await fn(item)
+            return res if is_flat_map else [res]
+
+    else:
+        raise ValueError(f"Expected a coroutine function, got {fn}")
+
+    # Goals of the algorithm applying async UDF application to the provided iterator
+    # are following:
+    #
+    #   - No more than `max_concurrency` async tasks are running
+    #     at any given moment
+    #   - Slow consumption from the output queue should result in
+    #     the processing to get back-pressured (so that output queue
+    #     doesn't grow unbounded)
+    #   - Order of the items (rows/batches) produced by this method
+    #     *must be* deterministic (though is not guaranteed to be specified
+    #     if max_concurrency > 1)
+    #
+    # To achieve that, algorithm applying async UDF to elements of the provided sequence
+    # is structured like following:
+    #
+    #   - Task scheduling and subsequent results re-ordering are performed as
+    #     different stages (inside `_schedule` and `_report` methods respectively)
+    #
+    #   - Scheduling stage aim to schedule and run no more than `max_concurrency` tasks
+    #     at any given moment
+    #
+    #   - Once task completes it's added into task completion queue for its results to be
+    #     subsequently reported with deterministic ordering). Task completion queue is
+    #     capped at `maxsize=max_concurrency` elements to make sure scheduling stage is
+    #     throttled (and task completion queue isn't growing unbounded) in case when
+    #     reporting stage isn't able to keep up.
+    #
+    #   - Reporting stage dequeues completed tasks from completion queue, reorders
+    #     them (to *always* produce deterministic ordering) and adds its results into
+    #     output queue.
+    #
+    #   - Output queue is capped at `maxsize=max_concurrency` elements to make sure that
+    #     reporting stage is throttled (and output queue doesn't grow unbounded) in case
+    #     when consumer (Ray task itself) isn't able to keep up
+    #
+    async def _schedule(it: Iterator[T], completed_tasks_queue: asyncio.Queue) -> None:
+        loop = asyncio.get_running_loop()
```

Comment:
Nit, explicitly use the udf_map_asyncio_loop, or pass it as an arg from _transform

Reply:
I was thinking the other way makes it more explicit that it will be using `udf_map_asyncio_loop`. No strong preference though.